lua/plenary/async_lib/work.lua

```diff
@@ -0,0 +1,91 @@
+local a = require('plenary.async_lib.async')
+local uv = vim.loop
+
+local M = {}
+
+local function work_map(tbl, func, callback)
+  local results = {}
+  local ran_callback = false
+
+  local work = uv.new_work(func, function(idx, ...)
+    results[idx] = {...}
+    if #results == #tbl and ran_callback == false then
+      callback(results)
+      ran_callback = true
+    end
+  end)
+
+  for idx, v in ipairs(tbl) do
+    M.validate_threadargs(v)
+
+    work:queue(idx, v)
+  end
+end
+
+M.map_async = a.wrap(work_map, 3)
+
+function M.is_threadarg(t)
+  return type(t) == "nil" or type(t) == "boolean" or type(t) == "number" or type(t) == "string" or type(t) == "userdata"
+end
+
+function M.validate_threadargs(...)
+  for arg in ipairs({...}) do
+    if not M.is_threadarg(arg) then
+      error('Was not a threadarg')
+    end
+  end
+end
+
+--- wrap function to execute in threadpool.
+--- need to figure out what work does
+--- this functions wraps it for a callback api
+--- WARNING: func cannot have any upvalues, only lua standard library
+--- WARNING: func can only be passed threadargs, see above
```

Comment:
it's actually possible to check some of these items with `debug.*` things if you're interested.

Reply:
I'll probably move work for a later pr as it is too experimental. We probably need to figure out how to set stuff in the new lua state so we can access neovim api functions anyway. there is a related [pr](https://github.com/luvit/luv/issues/531)